Configuring ICEUpdater to use Asio

amccaskey opened the pull request:
Configured Updater to use INetworkingTool and INetworkingToolFactory. Added Asio library to contrib. Added LibcurlUtils and AsioNetworkingTool as implementations of INetworkingTool, set AsioNetworkingTool as default implementation. Removed dependence on TBB threads and added dependence on CXX11. Commit messages below:

Setup Updater to use INetworkingTool with Asio
and Libcurl implementation. Post with Asio still not working correctly
though. (#5950)

Finally got http posts working with asio library (#5950)

Removed dependence on TBB since Asio and now Updater can only be used
with C++11 support (#5950).

Got everything working with both Asio and Curl, updated build to look for C++11 and add ASIO_STANDALONE flag, updated tests and they pass in serial and parallel (#5950)

Improved error handling for bad socket connections (#5950)

Updated build and tests to work with CXX11 is not enabled

Fixed some final bugs, updated response read to be more robust.

setting up proper moose syntax, ran delete trailing whitespace

Commit:
Adding ASIO to contrib.

Setup Updater to use INetworkingTool with Asio
and Libcurl implementation. Post with Asio still not working correctly
though. (#5950)

Finally got http posts working with asio library (#5950)

Removed dependence on TBB since Asio and now Updater can only be used
with C++11 support (#5950).

Got everything working with both Asio and Curl, updated build to look for C++11 and add ASIO_STANDALONE flag, updated tests and they pass in serial and parallel (#5950)

Improved error handling for bad socket connections (#5950)

Updated build and tests to work with CXX11 is not enabled

Fixed some final bugs, updated response read to be more robust.

setting up proper moose syntax, ran delete trailing whitespace

Commit:
Adding ASIO to contrib.

Setup Updater to use INetworkingTool with Asio
and Libcurl implementation. Post with Asio still not working correctly
though. (#5950)

Finally got http posts working with asio library (#5950)

Removed dependence on TBB since Asio and now Updater can only be used
with C++11 support (#5950).

Got everything working with both Asio and Curl, updated build to look for C++11 and add ASIO_STANDALONE flag, updated tests and they pass in serial and parallel (#5950)

Improved error handling for bad socket connections (#5950)

Updated build and tests to work with CXX11 is not enabled

Fixed some final bugs, updated response read to be more robust.

setting up proper moose syntax, ran delete trailing whitespace

dschwen commented:
I wonder if there isn't a better way to add contribs like this one. Those are an awful lot of files. Why do we need to copy them into our repository? This looks like something we should make a submodule, or just download using a bootstrap script if the ICEUpdater is enabled.

permcody commented:
I agree, 13000 lines is a lot of new code (more than a 10% increase to the MOOSE code base not including modules or existing contribs for a single feature). Would you mind looking at an option to download this library in the ICEUpdate script?

amccaskey commented:
Yea sure, I'll look into that.

Commit:
Removing Asio from contrib. Setup ICE install

script to also pull down Asio to contrib and build Libmesh/Moose with --enable-cxx11. Setup build.mk to add ASIO_STANDALONE flag if asio has been pulled down. Modified AsioNetworkingTool to only build if ASIO_STANDALONE has been passed.

Setup ICE install script to also pull down Asio to contrib and build Libmesh/Moose with --enable-cxx11. Setup build.mk to add ASIO_STANDALONE flag if asio has been pulled down. Modified AsioNetworkingTool to only build if ASIO_STANDALONE has been passed.

running delete whitespace script

Running delete whitespace on util.py

Commit:
Removing Asio from contrib. Setup ICE install

script to also pull down Asio to contrib and build Libmesh/Moose with --enable-cxx11. Setup build.mk to add ASIO_STANDALONE flag if asio has been pulled down. Modified AsioNetworkingTool to only build if ASIO_STANDALONE has been passed.

Setup ICE install script to also pull down Asio to contrib and build Libmesh/Moose with --enable-cxx11. Setup build.mk to add ASIO_STANDALONE flag if asio has been pulled down. Modified AsioNetworkingTool to only build if ASIO_STANDALONE has been passed.

running delete whitespace script

Running delete whitespace on util.py

amccaskey commented:
Just to recap - these latest commits remove Asio from framework/contrib, and add a modified version of our ICE python install script to the scripts directory. This new script installs ICE, pulls down Asio to framework/contrib/asio, executes update_and_rebuild_libmesh.sh with --enable-cxx11, then builds Moose and Moose tests. The build.mk file was modified to add -DASIO_STANDALONE if the contrib/asio directory exists and if --enable-cxx11 has been passed. I also added ifdefs to AsioNetworkingTool to only be built if the asio directory is available in contrib. 

I updated the python test scripts to enable an asio parameter in the iceupdater/tests file so that the test will only be run if Asio has been pulled down to contrib.

permcody commented on framework/moose.mk:
```diff
@@ -18,6 +18,12 @@ pcre_deps      := $(patsubst %.cc, %.$(obj-suffix).d, $(pcre_srcfiles)) \
+#asio_DIR	:= $(FRAMEWORK_DIR)/contrib/asio
+
+
+#
```
Do you need these lines? Looks like they may be important.

amccaskey replied:
That can be deleted, put it in there early on and forgot to remove it.

permcody commented on framework/src/outputs/ICEUpdater.C:
```diff
@@ -27,10 +28,16 @@ InputParameters validParams<ICEUpdater>()
   params += AdvancedOutput<Output>::enableOutputTypes("postprocessor");
 
+  // Get an MooseEnum of the available 'execute_on' options
+  MooseEnum networkingTools("asio=0x01 curl=0x02");
```
This comment may be incorrect. Also you probably should just use counting integers here instead of bit field integers here unless you really plan to use bitwise operators.

permcody commented on python/TestHarness/util.py:
```diff
@@ -67,6 +67,10 @@
+  'cxx11' :        { 're_option' : r'#define\s+LIBMESH_HAVE_CXX11\s+(\d+)',
+                     'default'   : 'FALSE',
+                     'options'   : {'TRUE' : '1', 'FALSE' : '0'}
+                   },
```
@jwpeterson - Is this the best way to detect MOOSE with C++11 in the TestHarness? It looks ok, just wanted to make sure you think so too.

jwpeterson replied:
This is the only way I know of to do it...

permcody commented on framework/contrib/ice_updater/include/INetworkingTool.h:
```diff
@@ -0,0 +1,104 @@
+
+	virtual ~INetworkingTool() {
+	}
+
```
Since this code is all in contrib, it doesn't need to follow our coding standards, but it would be nice if it would just for consistently. We use 2 spaces for all indentation (except for initialization lists). Also curly braces should always be on their own lines and matching indentation between opening and closing pairs. 

See the rest of the guidelines here: http://mooseframework.org/wiki/CodeStandards/

permcody commented on framework/contrib/ice_updater/include/Updater.h:
```diff
@@ -333,7 +394,15 @@ class UpdaterThread: public tbb::task,
    */
-  tbb::task* execute();
+  void execute() {
+	  processThread = MooseSharedPointer<std::thread>(new std::thread(&UpdaterThread::runThreaded, this));
```
You don't necessarily have to use MooseSharedPointer if you require c++11. It's fine either way or you could use std::shared_pointer

permcody commented:
This is looking better. You'll probably want to re-enable the existing skipped tests too.

Commit:
Fixing code style for INetworkingTool, deleting old comment in moose.mk, and updating ICEUpdater networkingTool param to use int

Commit:
Fixing code style for INetworkingTool, deleting old comment in moose.mk, and updating ICEUpdater networkingTool param to use int

permcody commented on framework/build.mk:
```diff
@@ -57,6 +57,16 @@ ifneq (,$(findstring mpi,$(cxx_compiler)))
+ifneq ("$(wildcard $(ASIO_FILE))","")
+# If it is installed, make sure this is a 
+# build using C++11, if so, then turn on Asio Networking
+ifeq (,$(findstring "-std=gnu++11",$(libmesh_CXXFLAGS)))
```
@jwpeterson one more - Can we rely on this flag to tell us if c++11 is enabled? Probably not for clang, or other compilers

jwpeterson replied:
No we can't.  I think clang supports that flag but it also supports `-std=c++11` and possibly others.  No idea about Intel or other compilers, but this is not a good check in general.

permcody replied:
@amccaskey - you shouldn't need this. You should probably just use the existing HAVE C++11 define in libMesh. You can use it to create a new preprocessor constant if you wish, but this isn't going to work.

amccaskey replied:
@permcody @jwpeterson So you guys are saying that checking for -std=gnu+11 won't work for some of the other compilers? How do you check in the Makefile for the LIBMESH_HAVE_CXX11 constant?

permcody replied:
That value is determined by several configure tests (m4 scripts). Basically
we do it the old tried and true way. We compile several small test problems
and see if they work or not. What we find is that most compilers support a
subset of C++11 features unless you explicitly tell the compiler to obey
ANSI standards. It's tricky to get this behavior right so you don't want to
reinvent the wheel here.

On Thu, Dec 10, 2015 at 7:40 AM Alex McCaskey <notifications@github.com>
wrote:

> In framework/build.mk
> <https://github.com/idaholab/moose/pull/6037#discussion_r47233115>:
>
> > @@ -57,6 +57,16 @@ ifneq (,$(findstring mpi,$(cxx_compiler)))
> >  	cxx_compiler = $(shell $(libmesh_CXX) -show)
> >  endif
> >
> > +# Check that we have Asio installed.
> > +ASIO_FILE := $(MOOSE_DIR)/framework/contrib/asio/include/asio.hpp
> > +ifneq ("$(wildcard $(ASIO_FILE))","")
> > +# If it is installed, make sure this is a
> > +# build using C++11, if so, then turn on Asio Networking
> > +ifeq (,$(findstring "-std=gnu++11",$(libmesh_CXXFLAGS)))
>
> @permcody <https://github.com/permcody> @jwpeterson
> <https://github.com/jwpeterson> So you guys are saying that checking for
> -std=gnu+11 won't work for some of the other compilers? How do you check in
> the Makefile for the LIBMESH_HAVE_CXX11 constant?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6037/files#r47233115>.
>

amccaskey replied:
I'm not exactly sure how to get build.mk to get the flag from libmesh_config.h. Any way I could get some help?

jwpeterson replied:
I think I'd probably just use a "grep" one-liner to look in libmesh_config.h for LIBMESH_HAVE_CXX11.

Commit:
switching c++11 check to use libmesh_config.h

permcody commented:
@amccaskey - Hope you had a nice holiday. I think we were very close to merging this before the break but alas, it is in conflict again due to other changes in the framework. Do you have some time to resolve these conflicts and any remaining outstanding issues?  Thanks

permcody commented:
ping @amccaskey

permcody commented:
Thanks for working on this but we typically don't like to merge into our topic branches. It really creates a mess in the history to merge from devel to the topic branch then back to devel when we do our final merge. Instead, we prefer rebase. Just rebase this whole branch on devel and it should figure out that those commits from @dschwen are already in devel and filter them out. This also makes reviewing this PR much easier.

amccaskey commented:
I did ```git pull --rebase devel```, then resolved merge conflicts. Is that wrong?

permcody commented:
That's correct. What you don't want to do is type "git merge" at any time.
Do you remember what commands you typed after you resolved the merge
conflicts?

Normally you "git add" the files in conflict and then type git rebase
--continue. You don't actually commit the files in conflict during the
rebase. If you need help with this. I can assist.

On Tue, Jan 19, 2016 at 9:30 AM Alex McCaskey <notifications@github.com>
wrote:

> I did git pull --rebase devel, then resolved merge conflicts. Is that
> wrong?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6037#issuecomment-172907804>.
>

amccaskey commented:
Yep I ran

```git pull --rebase devel```
Fixed the conflicts that arose in ICEUpdater, etc (mainly regarding the use of the have TBB preprocessor definition, which isn't needed anymore).
``` git add framework/``` To add the changed files,
``` git rebase --continue```

permcody commented:
That should have worked. Something went wrong though since Daniels commits
are still sitting on that branch. You can always try it again on another
branch. Again, if you still continue to have difficulties. I'll push your
branch up on your behalf.

On Tue, Jan 19, 2016 at 9:40 AM Alex McCaskey <notifications@github.com>
wrote:

> Yep I ran
>
> git pull --rebase devel
> Fixed the conflicts that arose in ICEUpdater, etc (mainly regarding the
> use of the have TBB preprocessor definition, which isn't needed anymore).
> git add framework/ To add the changed files,
> git rebase --continue
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6037#issuecomment-172910808>.
>

amccaskey commented:
@permcody tried it again, this one should be good (hopefully). Let me know if it needs anything else.

permcody commented:
Well it appears you managed to push your old branch back again. The one without the merge conflict fixes. I'm happy to push this up for you if you'd like. Just let me know.

amccaskey commented:
yes please do

permcody commented:
@amccaskey - for your records here are the commands I typed to pull, fix and repush your branch:

```
 1119  git fetch upstream
 1120  git fetch upstream pull/6037/head:alex
 1121  git co alex
 1122  git log
 1124  git diff
 1128  git rebase upstream/devel
 1129  git st
 1130  git diff
    # Fix the file in conflict here (In this case competing changes in Moose.C)
 1131  git add ../framework/src/base/Moose.C
 1132  git rebase --continue
 1133  git push origin
```

permcody commented:
Closing this PR - Moving to #6220
